Field probe reduced diagnostics

lgiacome opened the pull request:
This PR adds a new reduced diagnostics to probe the field at a given in point the domain.
It probes the 6 components of the EM field together with |E|**2 and |B|**2.
The fields can be probed as raw or cell-centered, while the energies are always cell-centered.

This feature is particularly useful to compute the resonant frequency of a cavity.

NeilZaim commented:
Hello @lgiacome ! Thanks a lot for adding this diag :+1:. I think it will be useful, especially since I think that we currently don't have access to the fields at the particles' position in the particle output.

I didn't have time to do a thorough review yet, and just had a quick look at the PR. I'm thinking that for now it may not work with multiple MPI processes. The main reason is that you're doing something like

```
const int i_probe = amrex::Math::floor((x_probe - prob_lo[0])/cell_size[0]);
const int j_probe = amrex::Math::floor((y_probe - prob_lo[1])/cell_size[1]);
const int k_probe = amrex::Math::floor((z_probe - prob_lo[2])/cell_size[2]);

for ( MFIter mfi(Ex, TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
    do stuff with arr_Ex(i_probe, j_probe, k_probe)
}
```

The MFIter loop is a loop over all the boxes in the domain, but only one of them contains the point `(i_probe, j_probe, k_probe)` (or maybe a few boxes contain the point if it's close enough to the border to be in the guard cells of another box). So if you try to access `arr_Ex(i_probe, j_probe, k_probe)` with an MPI process that does not contain this point, you can possibly segfault.

In any case, it would be a good idea to add automated test coverage for this reduced diag. The testing of the reduced diags is done [here](https://github.com/ECP-WarpX/WarpX/blob/development/Examples/Tests/reduced_diags/inputs) and [here](https://github.com/ECP-WarpX/WarpX/blob/development/Examples/Tests/reduced_diags/analysis_reduced_diags.py), you can have a look at how it's done and then add this new diag. Once the test is ready, you can try it locally with multiple MPI ranks and see if it works.

Also, I see that for now, you interpolate the field at the center of the cell that contains `(x_probe, y_probe, z_probe)`. I don't know what's best for the application you have in mind, but it's also possible to directly interpolate the field at the point `(x_probe, y_probe, z_probe)`, regardless of its position relative to the grid (that would be exactly like a field gathering).

If you have questions, feel free to ask them here or on slack. :)

lgiacome commented:
Hi @NeilZaim! Thanks a lot for your super useful comment, I will answer point by point.

> Hello @lgiacome ! Thanks a lot for adding this diag 👍. I think it will be useful, especially since I think that we currently don't have access to the fields at the particles' position in the particle output.
> 
> I didn't have time to do a thorough review yet, and just had a quick look at the PR. I'm thinking that for now it may not work with multiple MPI processes. The main reason is that you're doing something like
> 
> ```
> const int i_probe = amrex::Math::floor((x_probe - prob_lo[0])/cell_size[0]);
> const int j_probe = amrex::Math::floor((y_probe - prob_lo[1])/cell_size[1]);
> const int k_probe = amrex::Math::floor((z_probe - prob_lo[2])/cell_size[2]);
> 
> for ( MFIter mfi(Ex, TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
>     do stuff with arr_Ex(i_probe, j_probe, k_probe)
> }
> ```
> 
> The MFIter loop is a loop over all the boxes in the domain, but only one of them contains the point `(i_probe, j_probe, k_probe)` (or maybe a few boxes contain the point if it's close enough to the border to be in the guard cells of another box). So if you try to access `arr_Ex(i_probe, j_probe, k_probe)` with an MPI process that does not contain this point, you can possibly segfault.
> 

You were of course right. I tried the new field probe with 2 MPI ranks and with the measurement point assigned to the second rank and I got the segfault. 
I fixed this but I am not in love with my solution. I have put in place a weird way of communicating which processor contains (i_probe, j_probe, k_probe) by using a reduce max operation. This was needed to know which processor should communicate the quantities to output to the IO processor. Please let me know you have any suggestions on how to do this better.

> In any case, it would be a good idea to add automated test coverage for this reduced diag. The testing of the reduced diags is done [here](https://github.com/ECP-WarpX/WarpX/blob/development/Examples/Tests/reduced_diags/inputs) and [here](https://github.com/ECP-WarpX/WarpX/blob/development/Examples/Tests/reduced_diags/analysis_reduced_diags.py), you can have a look at how it's done and then add this new diag. Once the test is ready, you can try it locally with multiple MPI ranks and see if it works.
> 

I fully agree that it's a good idea to have also this reduced diag covered by the test. I added it to the test you pointed out and I also tried locally with multiple MPI ranks and now it works nicely.

> Also, I see that for now, you interpolate the field at the center of the cell that contains `(x_probe, y_probe, z_probe)`. I don't know what's best for the application you have in mind, but it's also possible to directly interpolate the field at the point `(x_probe, y_probe, z_probe)`, regardless of its position relative to the grid (that would be exactly like a field gathering).
> 

For the moment I didn't think of it that way, I was just thinking of saving raw data. For consistency with other diagnostics, I decided to optionally save the raw data while saving the averaged data by default. 
If interpolating to any point is something interesting for other use cases, I could do something like this;

- if the raw_fields option is set to true I "snap to the grid" and save the raw data for the cell containing the measurement point
- if the raw_fields option is set to false then I interpolate to the specified point doing the sort of gather you are suggesting

This should work well for everyone :)

If this is ok I can go ahead fixing the implementation. 
Just one question: for the field gather I would use the last `doGatherShapeN` function from `FieldGather.H` in the same way as it is done in `PhysicalParticleContainer.cpp` at line 1977, correct?

I will mark the PR again as WIP until I'm done with the new interpolation.

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,250 @@
+        amrex::IndexType const Bytype = By.ixType();
+        amrex::IndexType const Bztype = Bz.ixType();
+
+        amrex::Real hv_Ex, hv_Ey, hv_Ez, hv_E, hv_Bx, hv_By, hv_Bz, hv_B;
```
Here I think that hv initially stood for "highest value" which is not entirely accurate here. Feel free to find a new name if you want (it's also ok to leave it as it is).

lgiacome replied:
Ah of course.. I copy-pasted a lot frome the max field diagnostics, which explains these names.. :)
I changed the hv_ prefixes to fp_ (field probe)

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,250 @@
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+        for (amrex::MFIter mfi(Ex, TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+            const amrex::Box& box = mfi.validbox();
```
```suggestion
            const amrex::Box& box = amrex::enclosedCells(mfi.nodaltilebox());
```

I'm wondering if we should do this? Otherwise, I'm thinking that there may be a possibility, depending on the staggering, that the box contains the point `(i_probe, j_probe, k_probe)` although the position `(x_probe, y_probe, z_probe)` is in the guard cells. In which case, it's possible that there could be not enough points to perform the interpolation with order `WarpX::nox`.

lgiacome replied:
Sounds fair! I've changed as you suggested.

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,250 @@
+                                Bx_interp, By_interp, Bz_interp, arrEx, arrEy, arrEz, arrBx,
+                                arrBy, arrBz, Extype, Eytype, Eztype, Bxtype, Bytype, Bztype,
+                                dx_arr, xyzmin_arr, amrex::lbound(box),
+                                WarpX::n_rz_azimuthal_modes, WarpX::nox, false);
```
You don't necessarily have to use the same interpolation order as for the field gathering `WarpX::nox` but it's fine as well to use it. In any case, it could be good to specify what you choose to do in the parameters doc.

lgiacome replied:
Good point. To be as flexible as possible I added an optional parameter to specify the interpolation order, with 1 as the default value.

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,250 @@
+        amrex::ParallelDescriptor::ReduceIntMax(probe_proc);
+
+        if(amrex::ParallelDescriptor::MyProc()==probe_proc){
+            amrex::ParallelDescriptor::Send(&hv_Ex, 1,
```
I can imagine two scenarios where the MPI communications could deadlock:

1. The position `(x_probe, y_probe, z_probe)` is not in the physical domain. In this case, I'd expect `probe_proc` to remain -1 and the IO proc will probably wait indefinitely in the recv for a message to arrive. You could add a check that the position is inside the simulation domain before the MFIter loop and MPI communications, and if it is not you could fill the data arrays with zeros and return.
2. The probe_proc is the same as the IO proc. Since you're using blocking MPI calls, this could deadlock in theory (although in practice the exchanged data is small so it'll probably be bufferized). You can possibly skip the communications altogether if the probe_proc is the IO_proc. (using a non blocking send, which I think is Asend in amrex, should work as well)

lgiacome replied:
Both are fair points I solved them as follows:

1.  I am now checking if `(x_probe, y_probe, z_probe)` is inside the physical domain. If not I fill the data arrays with zeros.
2. I am skipping the communication if `probe_proc` is the same as the IO_proc

NeilZaim reviewed (commented):
Thanks for the udpates @lgiacome. I think your solution should work and I don't see anything that is much simpler. For now I only had a quick look at the `ComputeDiags` function and have some comments about some corner cases that could occur (see below). Unfortunately, I'll be on vacations next week so I won't have time to do a more thorough review before the week after.

ax3l commented:
@lgiacome did you have a chance to test this on GPU already? :)

NeilZaim commented on Docs/source/usage/parameters.rst:
```diff
@@ -1985,6 +1985,33 @@ Reduced Diagnostics
+    * ``FieldProbe``
+        This type computes the value of each component of the electric and magnetic fields
+        and of the norm of the electric and magnetic field vectors at a point in the domain.
+        The point where the fields are measured is specified through the inputs parameters
```
```suggestion
        The point where the fields are measured is specified through the input parameters
```

NeilZaim commented on Examples/Tests/reduced_diags/analysis_reduced_diags.py:
```diff
@@ -178,6 +178,14 @@
+values_yt['Bx(0.5, 0.5, 0.5)'] = Bx[24, 24, 24]
+values_yt['By(0.5, 0.5, 0.5)'] = By[24, 24, 24]
+values_yt['Bz(0.5, 0.5, 0.5)'] = Bz[24, 24, 24]
+values_yt['|B(0.5, 0.5, 0.5)|'] = np.sqrt(Bx[24, 24, 24]**2 + By[24, 24, 24]**2 + Bz[24, 24, 24]**2)
```
The name `(0.5, 0.5, 0.5)` is slightly misleading since the position of the probe is not 0.5 but 0.53. Not a big issue but it could be `Ex at probe` to be more general.

Also you could potentially replace the `[24,24,24]` with a variable so that it's easier to change the probe position.

NeilZaim commented on Examples/Tests/reduced_diags/analysis_reduced_diags.py:
```diff
@@ -250,6 +259,14 @@
+values_rd['Bx(0.5, 0.5, 0.5)'] = FPdata[1][6]
+values_rd['By(0.5, 0.5, 0.5)'] = FPdata[1][7]
+values_rd['Bz(0.5, 0.5, 0.5)'] = FPdata[1][8]
+values_rd['|B(0.5, 0.5, 0.5)|'] = FPdata[1][9]
```
Interesting that the test pass, even though we compare values interpolated to the probe position and values at the cell center (in the plotfiles) to 1e-12 relative accuracy. Does that work because the value of the probe position is chosen to be exactly at a cell center? If so, that could be specified somewhere in a comment (either in the input file or in the analysis file).

lgiacome replied:
Yes, I placed the probe at the cell center to match as closely as possible the value in the plotfiles. I'll comment on that in the input file.

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.H:
```diff
@@ -0,0 +1,53 @@
+/* Copyright 2020 Neil Zaim, Yinjian Zhao
```
Feel free to put your name there as well :)

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.H:
```diff
@@ -0,0 +1,53 @@
+
+};
+
+#endif // WARPX_DIAGNOSTICS_REDUCEDDIAGS_FIELDMAXIMUM_H_
```
```suggestion
#endif // WARPX_DIAGNOSTICS_REDUCEDDIAGS_FIELDPROBE_H_
```

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+#include "FieldProbe.H"
+#include "Particles/Gather/FieldGather.H"
+
+#include "Utils/CoarsenIO.H"
```
It's likely that we no longer need this line since we use the particle FieldGather do to the interpolation.

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.H:
```diff
@@ -0,0 +1,53 @@
+    static constexpr int index_absB = 7;
+
+private:
+    amrex::ParticleReal x_probe, y_probe, z_probe;
```
I would put `Real` instead of `ParticleReal`. (these are field quantities and not particle quantities)

NeilZaim replied:
Also, in 2D `y_probe` is probably not needed so it may be safer to not declare it at all.

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+    nLevel += 1;
+
+    amrex::ParmParse pp_rd_name(rd_name);
+    pp_rd_name.query("x_probe", x_probe);
```
I think this should be a `get` instead of a `query` (i.e. the code should stop if the value is not provided). In fact, it should probably be a `getWithParser` so that we can use the parser to define the probe position.

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+    pp_rd_name.query("y_probe", y_probe);
+    pp_rd_name.query("z_probe", z_probe);
+
+    const bool raw_specified = pp_rd_name.query("raw_fields", raw_fields);
+    if(!raw_specified) raw_fields = false;
```
To define default values, we typically do something like
```
raw_fields = false;
pp_rd_name.query("raw_fields", raw_fields);
```
(or possibly initializing `raw_fields` to `false` in `FieldProbe.H`), which saves an extra variable (`raw_specified` in your case). But what you do works as well, and I don't have a particularly strong opinion on this.

lgiacome replied:
I like your version better. I'll change it :)

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+    const bool raw_specified = pp_rd_name.query("raw_fields", raw_fields);
+    if(!raw_specified) raw_fields = false;
+
+    const bool interp_order_specified = pp_rd_name.query("raw_fields", interp_order);
```
```suggestion
    const bool interp_order_specified = pp_rd_name.query("interp_order", interp_order);
```

Same comment as above also applies

NeilZaim replied:
Also, to be extra safe, we could add a check that the selected interpolation order is not too large (otherwise segfaults or other weird things could happen). I think we could check that the allocated number of guard cells is at least equal to ceil(interp_order//2) (this formula should be verified independently :p) or more simply that interp_order <= WarpX::nox (more constraining).

lgiacome replied:
To keep it simple I'll go for the second one

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+}
+// end constructor
+
+// function that computes maximum field values
```
```suggestion
// function that computes field values at probe position
```

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+        const auto prob_lo = gm.ProbLo();
+        const auto prob_hi = gm.ProbHi();
+
+        const bool probe_in_domain = x_probe >= prob_lo[0] and y_probe >= prob_lo[1] and
+                                     z_probe >= prob_lo[2] and x_probe <= prob_hi[0] and
+                                     y_probe <= prob_hi[1] and z_probe <= prob_hi[2];
```
This probably does not work in 2D, where `z_probe` should be compared to `probe_lo/hi[1]` (`probe_lo/hi[2]` does not exist) and `y_probe` is irrelevant (probably safer to not declare it at all). You can have a look at how it's done in `FieldReduction.H` for an example.

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+            const auto cell_size = gm.CellSizeArray();
+
+            const int i_probe = amrex::Math::floor((x_probe - prob_lo[0]) / cell_size[0]);
+            const int j_probe = amrex::Math::floor((y_probe - prob_lo[1]) / cell_size[1]);
```
Here we should use `z_probe` in 2D (this convention is definitely error prone :D )

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+
+            const int i_probe = amrex::Math::floor((x_probe - prob_lo[0]) / cell_size[0]);
+            const int j_probe = amrex::Math::floor((y_probe - prob_lo[1]) / cell_size[1]);
+            const int k_probe = amrex::Math::floor((z_probe - prob_lo[2]) / cell_size[2]);
```
Here, `k_probe` should be set to 0 in 2D.

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+            const amrex::MultiFab &By = warpx.getBfield(lev, 1);
+            const amrex::MultiFab &Bz = warpx.getBfield(lev, 2);
+
+            // Prepare interpolation of field components to cell center
```
```suggestion
            // Prepare interpolation of field components to probe position
```

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+            const amrex::MultiFab &Bz = warpx.getBfield(lev, 2);
+
+            // Prepare interpolation of field components to cell center
+            // The arrays below store the index type (staggering) of each MultiFab, with the third
+            // component set to zero in the two-dimensional case.
```
```suggestion
            // The arrays below store the index type (staggering) of each MultiFab.
```

Since you're using an `IndexType` here, there should be no third component in 2D.

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+            amrex::IndexType const Bztype = Bz.ixType();
+
+            int probe_proc = -1;
+            // MFIter loop to interpolate fields to cell center and get maximum values
```
```suggestion
            // MFIter loop to interpolate fields to probe position
```

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+                    const auto &arrBy = By[mfi].array();
+                    const auto &arrBz = Bz[mfi].array();
+
+                    amrex::ParticleReal Ex_interp = 0._rt, Ey_interp = 0._rt, Ez_interp = 0._rt;
+                    amrex::ParticleReal Bx_interp = 0._rt, By_interp = 0._rt, Bz_interp = 0._rt;
```
```suggestion
                    amrex::Real Ex_interp = 0._rt, Ey_interp = 0._rt, Ez_interp = 0._rt;
                    amrex::Real Bx_interp = 0._rt, By_interp = 0._rt, Bz_interp = 0._rt;
```

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+                    amrex::ParticleReal Ex_interp = 0._rt, Ey_interp = 0._rt, Ez_interp = 0._rt;
+                    amrex::ParticleReal Bx_interp = 0._rt, By_interp = 0._rt, Bz_interp = 0._rt;
+
+                    amrex::Array<amrex::Real, 3> v_galilean = {{0}};
```
Make this const?

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+                        &xyzmin = WarpX::GetInstance().LowerCornerWithGalilean(box, v_galilean,
+                                                                               lev);
+                    const std::array<Real, 3> &dx = WarpX::CellSize(lev);
+                    amrex::GpuArray<amrex::Real, 3> dx_arr = {dx[0], dx[1], dx[2]};
+                    amrex::GpuArray<amrex::Real, 3> xyzmin_arr = {xyzmin[0], xyzmin[1], xyzmin[2]};
```
Make these const?

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+                    amrex::GpuArray<amrex::Real, 3> xyzmin_arr = {xyzmin[0], xyzmin[1], xyzmin[2]};
+
+                    //Interpolating to the probe point
+                    doGatherShapeN(x_probe, y_probe, z_probe, Ex_interp, Ey_interp, Ez_interp,
```
In 2d, `y_probe` could be declared and set to 0. immediatly before this call. Alternately, it could be declared and set to 0. in the header file (and it would be unused after except here).

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,280 @@
+                    amrex::ParallelDescriptor::Send(&fp_Ex,
+                                                    1,
+                                                    amrex::ParallelDescriptor::IOProcessorNumber(),
+                                                    0);
```
You can potentially use `index_Ex`, etc. as tags in the MPI calls.

lgiacome replied:
Good idea!

NeilZaim reviewed (commented):
Thanks for updating the PR. Although I have put a lot of comments, they are mostly minor. The most important ones are about making the diag also work in 2D, which I think is not the case for now.

lgiacome commented:
> @lgiacome did you have a chance to test this on GPU already? :)

Yes :)

lgiacome commented:
@NeilZaim Thanks a lot for all the suggestions. I think I have implemented all the changes. Let me know if there's anything else that looks odd to you :)

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,293 @@
+
+                    //Interpolating to the probe point
+#if (AMREX_SPACEDIM == 2)
+                    amrex::Real y_probe = 0;
```
```suggestion
                    constexpr amrex::Real y_probe = 0._rt;
```

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,293 @@
+            const auto cell_size = gm.CellSizeArray();
+
+#if (AMREX_SPACEDIM == 2)
+            const int i_probe = amrex::Math::floor((x_probe - prob_lo[0]) / cell_size[0]);
```
This line is the same in 2D and 3D so it can be potentially moved out of the `#if` in case you want to make it more compact. But it's already very clear right now so it's fine to leave it as it is.

lgiacome replied:
I like the idea of saving a whole line :D

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,293 @@
+                                                    amrex::ParallelDescriptor::IOProcessorNumber(),
+                                                    index_absB);
+                }
+                if (amrex::ParallelDescriptor::MyProc()
+                    == amrex::ParallelDescriptor::IOProcessorNumber()) {
+                    amrex::ParallelDescriptor::Recv(&fp_Ex, 1, probe_proc, index_Ex);
```
I'm wondering if we should also add a check that `probe_proc` is not equal to `-1`. Just in case the scenario where `probe_in_domain` is `true` but no box actually contains the probe can occur (which I'm not sure is possible, could this happen with EB? or with mesh refinement?).

lgiacome replied:
Probably it's just safer to check also that. I'm not sure `probe_proc` can be `-1` with EB but it's likely that it can happen with mesh refinement.

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,293 @@
+                                     y_probe >= prob_lo[1] and y_probe <= prob_hi[1] and
+                                     z_probe >= prob_lo[2] and z_probe <= prob_hi[2];
+#endif
+        amrex::Real fp_Ex, fp_Ey, fp_Ez, fp_E, fp_Bx, fp_By, fp_Bz, fp_B;
```
If you decide to have both the checks for `probe_in_domain` and `probe_proc != -1`, you could initialize these values to 0._rt, and then you don't have to do anything more in case one of these checks fail.

lgiacome replied:
Now I'm actually wondering if I should have an assert for both these two cases so that the code crashes if the user specifies an invalid position for the probe. What do you think?

NeilZaim replied:
Hmm, I'm not sure. Maybe I would leave it as it is (return 0s) because if we use this probe with a moving window, it's possible that the probe is inside the domain for some timesteps and outside for others.

lgiacome replied:
Fair point!

NeilZaim reviewed (commented):
Hello @lgiacome and thanks a lot for the updates. The PR looks almost good to me, see few minor comments below.

lgiacome commented:
@NeilZaim thanks for the new suggestions, I should have fixed everything. Let's see if the PR still passes all the checks though :)

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,283 @@
+                                                    index_absB);
+                }
+                if (amrex::ParallelDescriptor::MyProc()
+                    == amrex::ParallelDescriptor::IOProcessorNumber() and probe_proc != -1) {
```
The check for `probe_proc != -1` here is a bit superfluous since we already have a check for this above.

lgiacome replied:
Yep true :)

NeilZaim reviewed (approved):
Thanks! The tests are passing so I'll approve the PR (I have a comment below, but it's detail). Thanks again for adding this feature!

ax3l commented on Source/Diagnostics/ReducedDiags/FieldProbe.cpp:
```diff
@@ -0,0 +1,283 @@
+                    amrex::Real Bx_interp = 0._rt, By_interp = 0._rt, Bz_interp = 0._rt;
+
+                    const amrex::Array<amrex::Real, 3> v_galilean = {{0}};
+                    const auto
+                        &xyzmin = WarpX::GetInstance().LowerCornerWithGalilean(box, v_galilean,
+                                                                               lev);
```
Oh, this causes a compile error in single precision:
AppleClang:
```
Source/Diagnostics/ReducedDiags/FieldProbe.cpp:183:85: error: no viable conversion from 'const amrex::Array<amrex::Real, 3>' (aka 'const array<double, 3UL>') to 'const amrex::Vector<amrex::Real>' (aka 'const Vector<double>')
                        &xyzmin = WarpX::GetInstance().LowerCornerWithGalilean(box, v_galilean,
                                                                                    ^~~~~~~~~~
```

NVCC:
```
Source/Diagnostics/ReducedDiags/FieldProbe.cpp(183): error #312: no suitable user-defined conversion from "const amrex::Array<amrex::Real, 3UL>" to "const amrex::Vector<amrex::Real, std::allocator<amrex::Real>>" exists
```

ax3l replied:
-> #2409

lgiacome replied:
Thanks @ax3l I really didn't see that coming. I think @RemiLehe already fixed this bug with #2409, correct?
